### ISSUE TITLE
Request for new database: Clickhouse

### ISSUE THREAD

**Issue by movy** [dbcodeio/public#111]
Hello,
Your product looks very promising, especially for Clickhouse, as there're no good visual tools right now available for Clickhouse, even its own cloud is kinda half-baked and very pricey just for UI alone. Would appreciate if Clickhouse support is added to DBcode, cheers.

**Comment by v-e-y** [comment]
Hi @mikeburgh 
First of all, thanks for an awesome plugin! I recently purchased it, and I really appreciate the features it provides.
However, I noticed that ClickHouse is not currently supported. Since I work with ClickHouse extensively, it would be fantastic to have it integrated into the plugin. I saw that it’s in your roadmap, so I just wanted to express my interest and support for this feature.

**Comment by mikeburgh** [comment]
Thanks for the purchase!

I had actually made some progress on Clickhouse but had to re work how array's are treated.. so it got delayed awhile. That's done now so I should be able to get the core stuff in (like Postgres it has a lot of other features which I will add as people request them).

<img width="366" alt="Image" src="https://github.com/user-attachments/assets/2f0a958c-2e02-4306-9e3e-ca16f356a93e" />

For now I am using the HTTP protocol, the docs seems to indicate that's the prefer method (it's how official Node drivers work).. I figured that was better than using the Postgres or MySQL compatability as that seems to require additional configuration ?

**Comment by movy** [comment]
Heavy ClickHouse user here.
It is absolutely better to use ClickHouse protocol directly via official clickhouse-js driver, as different compatibility layers would miss certain clickhouse features, may not be always enabled on prod servers, and are not 100% compatible with all MySQL or Postgres dialect. 

More info:
https://clickhouse.com/docs/interfaces/mysql

From my experience, clickhouse-js  is _extremely_ stable by now, I'm using it daily in dozen of projects. I think, to have a decent clickhouse interface right within VSCode would be a fantastic addition to DBCode, and I'd be happy to beta-test this integration.

**Comment by v-e-y** [comment]
> For now I am using the HTTP protocol, the docs seems to indicate that's the prefer method (it's how official Node drivers work).. I figured that was better than using the Postgres or MySQL compatability as that seems to require additional configuration ?

Thanks for the explanation! To be honest, I usually just connect to ClickHouse in the simplest way possible — I don’t really go deep into the setup or know much about the different protocols.

So as long as the plugin works out of the box (or with minimal setup), that’s totally fine **for me**.

**Comment by mikeburgh** [comment]
@movy thanks for the confirmation. I am using the official driver, and yes it seems very solid.

@v-e-y lets find out... 

I just shipped version 1.11.0 with support for ClickHouse as a preview.

A not complete, but close to, list of what's still a work in progress: 

Introspection:
- Get constraint information (currently only seems possible by passing the create statements)
- Get function parameters (again only seems possible by parsing create statement) also they are not bound to a DB, so show in all DB's
- Compound datatypes like Array, Tuple, etc are not mapping properly and will raise errors about being unknown. Need to resolve the display of them (eg Array(String) is different to string[] of other engines with arrays.
- Date datatypes are not showing their precision in the DB Explorer
- Enums

Results:
- Date time display WILL BE WRONG, need to resolve all the timezone data and local display formats (determine how driver resolves it with fields that have a TZ specified)
- All data is currently mapping as a string, need to properly map them based on types.
- Dynamic data types need attention
- numbers bigger than the JS number limit need to test to make sure we don't loose precision.
- Enums

Data Editing
- Works for simple columns, probably wont for arrays, tuples, and other compound ones.

Some questions
- Does it make sense to expose a role on the connection configuration ?
- Deleting data, it currently just uses delete, does it need to instead use ALTER TABLE DELETE
- What are some of the non standard features (Sort keys, partition keys for example) you would like to see surfaced in the introspection/explorer ?

**Comment by v-e-y** [comment]
Hi!
First of all — thank you so much for the release!
It feels like it already covers my needs — I can browse tables, run test queries, and generally work with my ClickHouse databases without any big issues. Really appreciate how quickly you put this preview together!

Background:
I’ve been working in a product company for over 2 years.
We have several projects, and we use ClickHouse everywhere — handling billions of records, clusters, and so on.

Here’s my feedback:
I started testing ClickHouse support, just connection and couple simple queries.

I have two groups: Dev and Prod.
When I add a ClickHouse connection, it connects successfully — everything looks good.
However, after opening a specific table from the created database, the green indicator disappears, and the status shows as disconnected.
At that point, I can’t move the connection to a group or edit it anymore.
These notifications appear
```txt
Only connections can be grouped.
Error connecting to prod_clickhouse-eu-2, error: getaddrinfo ENOTFOUND http
```
Despite that, I can still browse tables and run queries without any issues.
Restarting VSCode solves the problem.

Interestingly, if I move the connection to a group immediately after creating it — before connecting — everything works fine.

⸻

Regarding your additional questions:
	1.	As far as I know, we don’t use roles in our current ClickHouse setups.
	2.	In our case, we almost never delete data or columns inside ClickHouse. We usually delete entire tables or databases when needed, and that is handled by our DevOps team at a lower level.
	3.	I haven’t found anything important missing so far. Everything I personally need is already available.

⸻

Again, huge thanks for your work!
If you need more detailed feedback, I’m happy to help!

**Comment by v-e-y** [comment]
Yeah, I also noticed the array issue. But honestly, in our case, we only have one (not important) table using arrays — we found them a bit slow to work with.

**Comment by v-e-y** [comment]
I also found a problem when trying to use the **Contains** search on a column with type UUID.
It results in an error:
> Illegal type UUID of argument of function like: While processing id LIKE ‘%ce1b6b08-4483-37e3-b631%’.

In native/raw queries, we usually solve this by converting the UUID to a string first, like this:
```sql
WHERE toString(id) ILIKE '%ce1b6b08-4483-37e3-b631%';
```

**Comment by meguoe** [comment]
When executing a statement containing "EXPLAIN" and "WITH", the results cannot be printed correctly at present.

**Comment by Zelotiki** [comment]
Hi! I really like your extension for using SQL, but I ran into some problems with the native connection to Clickhouse from ports such as 9000, could you add this type of connection for the best experience using the extension?

**Comment by mikeburgh** [comment]
@Zelotiki the official clickhouse driver only supports the HTTP protocol, so port 9000 wont work. Is there a reason you prefer the native protocol over the http one ?

**Comment by Zelotiki** [comment]
@mikeburgh hey, Mike! The main reasons for us to use this port in :

- Higher speed – the native protocol is more efficient than HTTP.
- Less delays – no HTTP headers and no parsing.
- Compression support – reduces the amount of data transferred.
- Best compatibility – most customers use it.
- We use this protocol for our secure TLS channel, which provides us with access, unlike unsecured HTTP (I think this is the standard for corporate users)
[offical docs](https://clickhouse.com/docs/interfaces/formats/Native)

**Comment by mikeburgh** [comment]
@Zelotiki thanks for the information…. It looks like they have not provided any docs for the native protocol and I don’t think I am going to be attempting to reverse it anytime soon…
https://clickhouse.com/docs/interfaces/tcp

The official client mentions that it supports compression so I will look into options for it (I recall somewhere they mentioned it was not possible with read only) and it has https/ssl support https://clickhouse.com/blog/node-js-client-for-clickhouse

But I get it, corporate policies and what you prefer is what matters… if there is a way to get native protocol support to work in a reliable non reverse engineered way I will certainly look at it

**Comment by movy** [comment]
Compression is engaged via:
```js
createClient({
            url,
            compression: {
                response: true,
                request: true
            },
...
```

**Comment by mikeburgh** [comment]
@movy thanks, compression has been enabled in 1.11.9 which I just released.

**Comment by mikeburgh** [comment]
Closing this one off as done now.. Last few tweaks were made in the last releases.
